crates/holochain/src/core/present.rs

```diff
@@ -0,0 +1,196 @@
+//! # Present Data Searches
+//! Checking for data and metadata presence in different
+//! locations.
+//! Once data / metadata is found the type of dependency is
+//! recorded so that consumers of this api can make decisions
+//! about what type of dependency they require (e.g. waiting for
+//! a PendingValidation dep to validate).
+//! This is still a work in progress and will eventually
+//! replace the "present.rs" in the sys_validation_workflow.
+use holo_hash::{EntryHash, HeaderHash};
+use holochain_state::{
+    error::DatabaseResult,
+    fresh_reader,
+    prelude::JudgedPrefix,
+    prelude::{PendingPrefix, PrefixType},
+};
+use holochain_zome_types::element::{Element, SignedHeaderHashed};
+
+use crate::core::state::cascade::Cascade;
+
+pub use error::*;
+
+use super::{
+    state::element_buf::ElementBuf, state::metadata::MetadataBuf, state::metadata::MetadataBufT,
+    validation::Dependency,
+};
+
+mod error;
+
+macro_rules! found {
+    ($r:expr) => {
+        if let Some(r) = $r {
+            return Ok(Some(r));
+        }
+    };
+}
+
+/// A pair containing an element buf and metadata buf
+/// with the same prefix.
+/// This is useful for when you want to check
+/// a level for some data. For example
+/// you might want see if an entry exists in the
+/// "judged" element buf and the metadata buf.
+pub struct DbPair<'a, P, M>
+where
+    P: PrefixType,
+    M: MetadataBufT<P>,
+{
+    pub element: &'a ElementBuf<P>,
+    pub meta: &'a M,
+}
+
+/// A source of data that contains different levels.
+/// This trait allows different workspaces to provide
+/// the source of the data / metadata for the retrieve calls.
+/// You can think of this as a superset of the cascade where we want
+/// to be able to also check for data in our other local stores.
+// TODO: This might need to be broken into sub traits as
+// some workspaces don't have all the levels (e.g. IntegratedDhtOpsWorkspace
+// doesn't have pending).
+pub trait DataSource {
+    fn cascade(&mut self) -> Cascade;
+    fn pending(&self) -> DbPair<PendingPrefix, MetadataBuf<PendingPrefix>>;
+    fn judged(&self) -> DbPair<JudgedPrefix, MetadataBuf<JudgedPrefix>>;
+}
```

Comment:
I feel like we should start diagramming this stuff like we do for the workflows, and keep it up to date. I am not up to date on all the ins and outs of retrieval, and that would really help me. Maybe we can pair and/or work with Art to do that.

Reply:
I think you're right about this being a kind of extension to the cascade. Maybe we should look at incorporating it into the cascade so the cascade can be built with different constructors that gives it access to different sources or something.

What would make this easier is being able to iteratively work with these ideas over a few PRs. I think the data source type will probably go away but it's a good experiment to see what is actually needed if we do put it into the cascade.

We can also diagram it.  To me this whole problem comes down to the question of when can you trust data from somewhere when validating. A good example of this came up when writing this pr:
- Can I use the `ZomeId` from a **invalid** `CreateLink` header when deciding which zome to call for a `DeleteLink`?
  - If we can then what if the reason that the `CreateLink` was invalid was because the `ZomeId` was out of range?
  - If we can't then we must only check the in places that hold valid data (integrated, authored, cache) and places that might be valid (pending, judged) if we later check they are valid and 
  - If we can't then we can only get it from the network where the node serving the data has check the `ZomeId`. Which for a `CreateLink` is the `StoreElement` authority. However the sys validation doesn't actually check `ZomeId` on `CreateLink` only on the app entry type. It probably should but who the `RegisterLinkAdd` authority or the `StoreElement` authority or both.
  - If we can't then what if the `CreateLink` is invalid? Should the `DeleteLink` be then invalid as well? 

You can see how complex this gets very quickly and this is just one example of many such relationships. The simple thing to do is so we can trust **data** from invalid headers but this seems kind of odd to me. What happens to the `DeleteLink` that is "valid" be deletes an invalid `CreateLink` because it's `ZomeId` doesn't exist? Which app validation does it call?

The other point is in this example `DeleteLink` has a requirement on a valid `CreateLink` in sys validation. So do we just trust that fact and in app validation use raw retrieves that just get the data from anywhere? Seems kind of odd to even check the `rejected` store if we are counting on this only being valid. So maybe that leads to another type of call:
- "The only check places with valid (or potentially valid) data but don't worry about actually checking it's valid because we know that sys validation will do that" call hahaha.

Also this only makes any sense if `StoreElement` actually runs the `RegisterLinkAdd` validation which it doesn't.

Anyway that was a really long answer but I hope it gives you some insight into why I am playing with this stuff and not just modifying the cascade because I don't actually know what the right design is yet.

I think instead of diagrams we need tests that show examples like the above. A failing test is probably worth a thousand zoom calls :laughing:

Reply:
Yeah that is really complicated!! And it all seems to be necessary. What this indicates to me is that we don't yet have the right conceptual framework or *grammar* for expressing these relationships and dependencies. I think you are building the pieces of that grammar now in this PR and the last, but what's missing is a common sense, shared understanding of what that grammar actually is. The idea of having "levels of validation" is really good, as well as the different layers of persistence with the prefixes, those seem like the key components of the grammar. What's not super clear yet is *in general* how it all fits together.

This might be a case where there is no general statement that can be said, because every case is kind of exceptional in its own way, so you just have to understand each case. But even then, to have all the cases laid out in one easily-scannable place would let the eye take it all in and understand it better. That could take the form of a nice concise test suite, but we're not quite there yet, so that's why I suggested diagramming or some kind of non-code documentation.

I feel like there must be some unified way of describing all the requirements of validation in a tighter fashion, perhaps a more declarative approach, a language.

I'm kind of more concerned about this being understandable than about it being correct. That is the really hard problem, and I suspect it may take something along the lines of a small DSL to get into a manageable state. And I don't expect you to have built a DSL by now, and I don't see any way this could be simpler myself, so that's why I approved. Just chiming in with what I think needs to eventually happen.

Reply:
Yeh that's a good point. Perhaps we need a better way of capturing what follows a PR leaves behind. Especially when those follow ups don't fit in the parent story. So the context and clarity doesn't get lost later when we come back to it or worse we just never come back to it.
I don't know if backlog items really fit for this as it's more like several related follow ups need to be captured with diagrams and documentation. 
What if we literally iterate. So this is the app validation story (backlog item) right. Let's create app validation V2 now and start filling in the tasks from this discussion with some diagrams. We could do the same with sys validation. Then we know it's not done but we can draw a line on how much time we spend on it now. Thoughts?

Reply:
Sounds good, we should at least always track the next step in the process if there is a gap. And that next step might lead to next steps, but we only do one step at a time anyway. We could talk about it next sync.